datameta/storage.py

```diff
@@ -32,3 +43,94 @@ def rm(request, storage_path):
+        os.mkdir(outdir)
+    return os.path.join(outdir, db_file.storage_uri[7:]) # Strip the file:// prefix
+
+def write_file(request, db_file, file):
+    """Write the file content specified by 'file' to the storage denoted in 'db_file'"""
+    # Sanity checks and output path generation
+    if db_file.storage_uri is None or not db_file.storage_uri.startswith("file://"):
+        raise RuntimeError(f"Unable to store to storage URI '{db_file.storage_uri}'")
+    out_path = get_local_storage_path(request, db_file)
+
+    # Write the file
+    if not demo_mode(request):
+        file.seek(0)
+        with open(out_path, 'wb') as outfile:
+            shutil.copyfileobj(file, outfile)
+        log.info(f"[STORAGE][NEWFILE][user={db_file.user.uuid}][file={db_file.uuid}]")
+    else:
+        log.info(f"[!!DEMOMODE!!][STORAGE][NEWFILE][user={db_file.user.uuid}][file={db_file.uuid}]")
+
+def _freeze_local(request, db_file):
+    # Perform checksum comparison
+    try:
+        with open(get_local_storage_path(request, db_file), 'rb') as infile:
+            # Calculate filesize
+            infile.seek(0,2)
+            filesize = infile.tell()
+            infile.seek(0)
+            # Calculate checksum
+            md5 = hashlib.md5(infile.read()).hexdigest()
+        if md5 != db_file.checksum:
+            raise ChecksumMismatchError()
+        # Denote the filesize and mark the file as uploaded
+        db_file.filesize          = filesize
+        db_file.content_uploaded  = True
+    except FileNotFoundError:
+        raise NoDataError()
+
+def _freeze_s3(request, db_file):
+    # TODO
+    # - Move object to another S3 object unaccessible by presigned URL
+    # - Calculate MD5 for copied S3 object, if mismatch delete it and raise error
+    # - Update the storage_uri to the new S3 object
+    # - Delete old S3 object for which presigned URL was issued?
+    # - Mark file as uploaded
+    raise NotImplementedError()
+
+def freeze(request, db_file):
+    """Freezes a File. This function ensures that data is present, consistent
+    with the pre-announced checksum and if that is the case, calculates and
+    annotates the filesize and marks the file as content_uploaded
+
+    Args:
+        request - The calling HTTP request
+        db_file - The database 'File' object
+
+    Raises:
+        NoDataError - No data was uploaded for this file yet
+        NotWritableError - The file was already frozen
+        ChecksumMismatchError - The uploaded data does not match the pre-announced checksum
+    """
+    if db_file.storage_uri is None:
+        raise NoDataError() # No data has been uploaded yet
+    if db_file.content_uploaded:
+        raise NotWritableError() # Data has been uploaded and file was frozen already
+    if db_file.storage_uri.startswith("file://"):
+        return _freeze_local(request, db_file)
+    if db_file.storage_uri.startswith("s3://"):
+        return _freeze_s3(request, db_file)
+    raise NotImplementedError()
```

Comment:
I think for now that overall design for storage is fine.

However, for future maintainability, one might consider to restructure this in classes.
For instance, there is one base class `StorageHandlerBase` and the can be multiple specific implementations inheriting from that class. Each class represents an interface for interacting with a specific storage system (e.g. local FS and s3) and it implements system-specific functions for:
- retrieving storage paths for upload
- freezing a file
- (optionally) writing a file

Like this other users (who setup datameta) could also very simply provide their custom storage handler function (maybe for FTP or so).

Reply:
Currently, it also decides which storage backend to use by looking at the storage_uri of a specific file object. This might potentially be dangerous, since, if someone could hack a custom storage_uri into the database, he could access something he should not have access to.